src/navier.f90

```diff
@@ -1280,5 +1289,255 @@ subroutine tbl_flrt (ux1,uy1,uz1)
+       write(96,*) real((itime-1)*dt,mytype), qmm !write pressure drop
+    endif
+
+    !Correction
+    do j=1,xsize(2)
+        if (istret.eq.0) ym=real(j+xstart(2)-1-1,mytype)*dy-yc
+        if (istret.ne.0) ym=yp(j+xstart(2)-1)-yc
+        do k=1,xsize(3)
+            zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.ra.and.ep(i,j,k).eq.0) then
+                    ux(i,j,k)=ux(i,j,k)+(constant-qmm)
+                endif
+            enddo
+        enddo
+    enddo
+
+    !Check new bulk velocity
+    qmm     = zero
+    ncountt = zero
+    qm      = zero
+    ncount  = zero
+    do k=1,xsize(3)
+        zm=dz*real(xstart(3)-1+k-1,mytype)-zc
+        do j=1,xsize(2)
+            if (istret.eq.0) ym=real(j+xstart(2)-1-1,mytype)*dy-yc
+            if (istret.ne.0) ym=yp(j+xstart(2)-1)-yc
+            r=sqrt(ym*ym+zm*zm)
+            do i=1,xsize(1)
+                if (r.le.ra.and.ep(i,j,k).eq.0) then
+                    qm=qm+ux(i,j,k)
+                    ncount=ncount+one
+                else
+                    !Cancel solid zone (ra <= r <= ra+wt)
+                    !and buffer zone (r > ra+wt)
+                    ux(i,j,k)=zero
+                    uy(i,j,k)=zero
+                    uz(i,j,k)=zero
+                endif
+            enddo
+        enddo
+    enddo
+    call MPI_ALLREDUCE(qm,qmm,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    call MPI_ALLREDUCE(ncount,ncountt,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+    qmm=qmm/ncountt
+    if ((nrank==0).and.(mod(itime, ilist)==0)) print *,'    Bulk velocity  after',qmm
+    !
+    return
+    !
+  end subroutine pipe_bulk_u
+  !********************************************************************
+  !
+  subroutine pipe_bulk_phi(phi,ux,ep,is,constant)
+  !
+  !********************************************************************
+
+    use decomp_2d
+    use decomp_2d_poisson
+    use variables
+    use param
+    use ibm_param, only: ra
+    use MPI
+
+    implicit none
+    real(mytype),dimension(xsize(1),xsize(2),xsize(3))  :: phi,ux,ep
+    real(mytype)                                        :: constant      !bulk temperature value
+    real(mytype)                                        :: qv,qvm,qm,qmm !volumetric averaged values
+    real(mytype)                                        :: ym,zm,yc,zc,r
+    real(mytype)                                        :: ncount,ncountt
+    real(mytype)                                        :: phi_out
+    integer                                             :: ifile,is,j,i,k,code
+    character(len=30)                                   :: filename
+
+    if (iscalar.eq.0) return
+
+255 format('Tb_',I2.2,'.dat')
+256 format(' Scalar:                       #',I2)
+
+    ifile=50+is
+    if (itime.eq.ifirst.and.nrank.eq.0) then
+        write(filename,255) is
+        open(ifile,file=filename,status='unknown')
```

Comment:
Can we avoid fixed IO units ? Please add to the subroutine an array of integer variables with the save attribute to store the IO unit number.
https://github.com/xcompact3d/Incompact3d/blob/master/src/tools.f90#L1002

Reply:
Solved in https://github.com/xcompact3d/Incompact3d/pull/193/commits/e01c57e053605b919ec2a47187e25658c1c344b4